src/shared/extensions.ts

```diff
@@ -18,11 +18,29 @@ export const VSCODE_EXTENSION_ID = {
 /**
  * Long-lived, extension-scoped, shared globals.
  */
-export interface ExtContext extends vscode.ExtensionContext {
-    awsContext: AwsContext
-    regionProvider: RegionProvider
-    settings: SettingsConfiguration
-    outputChannel: vscode.OutputChannel
-    telemetryService: TelemetryService
-    chanLogger: ChannelLogger
+export class ExtContext implements vscode.ExtensionContext {
+    public asAbsolutePath(relativePath: string): string {
+        return this.vscodeContext.asAbsolutePath(relativePath)
+    }
+
+    public constructor(
+        private readonly vscodeContext: vscode.ExtensionContext,
+        public readonly awsContext: AwsContext,
+        public readonly regionProvider: RegionProvider,
+        public readonly settings: SettingsConfiguration,
+        public readonly outputChannel: vscode.OutputChannel,
+        public readonly telemetryService: TelemetryService,
+        public readonly chanLogger: ChannelLogger,
+
+        //
+        // Inherited properties:
+        //
+        readonly storagePath: string | undefined = vscodeContext.storagePath,
```

Comment:
What the code is doing with inherited properties in order to continue to expose what is available on vscode.ExtensionContext feels like doubling down on inheritance instead of switching over to composition. This seems like it sets ourselves up for problems in the future when the api surface increases.

It appears that the destructuring was the root cause of the problem. A ToolkitContext that doesn't extend/implement anything, which contains and exposes a vscode.ExtensionContext property, as well as any other properties we want (like aws context) seems like a more future proof approach.

Reply:
Feels questionable to create new "extension context" type which also contains an "extension context". But yes, the workaround here is also questionable :)

> It appears that the destructuring was the root cause of the problem.

Could it be a vscode bug? Or was there a technical problem with the way destructuring was used (if so , what was it)?

Reply:
Stable, boring code tends to be easier to explain, and has a lower likelihood of mystery bugs.

VS Code provides this extension context object to extensions. That object may or may not contain more data than what is typed in the version of `@types/vscode` that is used at any given time. The destructuring that took place [here](https://github.com/aws/aws-toolkit-vscode/pull/1081/files#diff-45327f86d4438556066de133327f4ca2L110) caused a bunch of getters to be called in order to produce the resulting object. It hit [this getter](https://github.com/microsoft/vscode/blob/84fc970a24933749e30c222054ec3f33c78c205e/src/vs/workbench/api/common/extHostExtensionService.ts#L380), which caused the exception that breaks extension activation.

Reply:
Is destructuring not considered stable/boring?